src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,806 @@
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+/**
+ * \brief Handles incoming IPC commands for Crossover component.
+ */
+static int crossover_cmd(struct comp_dev *dev, int cmd, void *data,
+			 int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+		comp_info(dev, "crossover_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = crossover_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = crossover_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	default:
+		comp_err(dev, "crossover_cmd(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+/**
+ * \brief Sets Crossover Filter component state.
+ * \param[in,out] dev Crossover Filter base component device.
+ * \param[in] cmd Command type.
+ * \return Error code.
+ */
+static int crossover_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "crossover_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+/**
+ * \brief Copies and processes stream data.
+ * \param[in,out] dev Crossover Filter base component device.
+ * \return Error code.
+ */
+static int crossover_copy(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sinks[SOF_CROSSOVER_MAX_STREAMS] = { NULL };
+	int i, ret;
+	uint32_t num_sinks = 0;
+	uint32_t frames = UINT_MAX;
+	uint32_t source_bytes, avail;
+	uint32_t flags = 0;
+	uint32_t sinks_bytes[SOF_CROSSOVER_MAX_STREAMS] = { 0 };
+
+	comp_dbg(dev, "crossover_copy()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 sink_list);
+
+	/* Check for changed configuration */
+	if (cd->config_new) {
+		crossover_free_config(&cd->config);
+		cd->config = cd->config_new;
+		cd->config_new = NULL;
+		ret = crossover_setup(cd, source->stream.channels);
+		if (ret < 0) {
+			comp_err(dev, "crossover_copy(), setup failed");
+			return -EINVAL;
+		}
+	}
+
+	/* Use the assign_sink array from the config to route
+	 * the output to the corresponding sinks
+	 */
+	num_sinks = crossover_assign_sinks(dev, cd->config, sinks);
+
+	buffer_lock(source, &flags);
+
+	/* check if source is active */
+	if (source->source->state != dev->state) {
+		buffer_unlock(source, flags);
+		return -EINVAL;
+	}
+
+	for (i = 0; i < SOF_CROSSOVER_MAX_STREAMS; i++) {
+		if (!sinks[i])
+			continue;
+		buffer_lock(sinks[i], &flags);
+		avail = audio_stream_avail_frames(&source->stream,
+						  &sinks[i]->stream);
+		frames = MIN(frames, avail); /* TODO does this work with an XRUN */
+		buffer_unlock(sinks[i], flags);
+	}
+
+	buffer_unlock(source, flags);
+
+	source_bytes = frames * audio_stream_frame_bytes(&source->stream);
+
+	for (i = 0; i < SOF_CROSSOVER_MAX_STREAMS; i++) {
+		if (!sinks[i])
+			continue;
+		sinks_bytes[i] = frames *
+				 audio_stream_frame_bytes(&sinks[i]->stream);
+	}
+
+	cd->crossover_process(dev, source, sinks, num_sinks, frames);
+
+	/* update components */
+	for (i = 0; i < SOF_CROSSOVER_MAX_STREAMS; i++) {
```

Comment:
We keep the number of sinks in the config. But the size of the sink array is initialized to SOF_CROSSOVER_MAX_STREAMS (4).

Maybe a better way to do this would be to initialize the sink array to be cd->config->num_sinks. We can probably safely do that since the config would have already been validated before being assigned. What do you think?

I just had it that way since mux does the same thing too.